apps/arweave/src/ar_merkle.erl

```diff
@@ -190,6 +196,112 @@ validate_path_strict_data_split(_ID, _Dest, _LeftBound, _RightBound, _Path, _Pat
 		_DataSize) ->
 	false.
 
+%% @doc Validate the given merkle path where any subtrees may have 0-based offset.
+validate_path2(ID, Dest, RightBound, _Path) when RightBound =< 0 ->
+	?LOG_ERROR([{event, validate_path2_called_with_not_positive_right_bound},
+			{root, ar_util:encode(ID)}, {dest, Dest}, {right_bound, RightBound}]),
+	throw(invalid_right_bound);
+validate_path2(ID, Dest, RightBound, Path) when Dest >= RightBound ->
+	validate_path2(ID, RightBound - 1, RightBound, Path);
+validate_path2(ID, Dest, RightBound, Path) when Dest < 0 ->
+	validate_path2(ID, 0, RightBound, Path);
+validate_path2(ID, Dest, RightBound, Path) ->
+	validate_path2(ID, Dest, 0, RightBound, Path, byte_size(Path), RightBound, 0).
+
+validate_path2(ID, _Dest, LeftBound, RightBound,
+		<< Data:?HASH_SIZE/binary, EndOffset:(?NOTE_SIZE*8) >>, PathSize, DataSize,
+		LeftBoundShift) ->
+	case EndOffset - LeftBound > ?DATA_CHUNK_SIZE
+			orelse RightBound - LeftBound > ?DATA_CHUNK_SIZE of
+		true ->
+			io:format("Bounds: E=~B L=~B R=~B~n", [EndOffset, LeftBound, RightBound]), % TODO
+			false;
+		false ->
+			ChunkSize = EndOffset - LeftBound,
+			ValidateSplit =
+				case validate_strict_split of
+					_ when ChunkSize == (?DATA_CHUNK_SIZE) ->
+						LeftBound rem (?DATA_CHUNK_SIZE) == 0;
+					_ when EndOffset == DataSize ->
+						Border = RightBound - RightBound rem (?DATA_CHUNK_SIZE),
+						RightBound rem (?DATA_CHUNK_SIZE) > 0
+								andalso LeftBound =< Border;
+					_ when PathSize > ChunkSize ->
+						false;
+					_ ->
+						LeftBound rem (?DATA_CHUNK_SIZE) == 0
+								andalso DataSize - LeftBound > (?DATA_CHUNK_SIZE)
+								andalso DataSize - LeftBound < 2 * (?DATA_CHUNK_SIZE)
+				end,
+			case ValidateSplit of
+				false ->
+					io:format("Split: E=~B L=~B R=~B C=~B~n",
+							[EndOffset, LeftBound, RightBound, ChunkSize]), % TODO
+					false;
+				true ->
+					case hash([hash(Data), hash(note_to_binary(EndOffset))]) of
```

Comment:
is it possible to refactor this with the non-rebased check [here](https://github.com/ArweaveTeam/arweave/pull/433/files#diff-da5e1b73bd29394af0e871a3dc4b2299cafb3a0cc8aeb5b072f362667865918aL53-L56)?

Reply:
Actually... is there some way to combine both the `validate_path` and `validate_path2` functions into a single function? E.g. by setting the rebase to 0 or something when calling `validate_path`? Or maybe it's LeftBoundShift that needs to be set when validating a non-rebased tree?